engines/registry/app/services/registry_catalog_service.rb

```diff
@@ -0,0 +1,68 @@
+require 'json'
+require 'net/http'
+
+CATALOG_API_URL = 'http://127.0.0.1:5000/v2/_catalog'.freeze
+AUTH_URL = 'https://smt-ec2.susecloud.net/api/registry/authorize'.freeze
+CATALOG_SCOPE = 'registry:catalog:*'.freeze
+SERVICE = 'SUSE Linux OCI Registry'.freeze
+
+class RegistryCatalogService
+  attr_accessor :catalog_api_url
+
+  # We overwrite the public request /v2/_catalog with our endpoint
+  # RMT still needs to be able to get the unfiltered _catalog from the registry
+  # So we query CATALOG_API_URL (registry catalog) and return the fetched repositories
+  # n > 1000 crashes the request to CATALOG_API_URL
+
+  def initialize(url = CATALOG_API_URL)
+    @catalog_api_url = url
+  end
+
+  # be aware that this takes about 20-25 seconds to be finished if not in cache
+  def repos(reload: false, system: nil)
+    Rails.cache.fetch(@catalog_api_url, expires_in: 1.hour, force: reload) do
+      fetch_registry_repos(system)
+    end
+  end
+
+  private
+
+  def fetch_registry_repos(system)
+    Rails.logger.info('Fetch registry repos')
+    response = catalog_token(system)
+    catalog_auth_token = JSON.parse(response.body).fetch('token', '')
+    response = all_repos(catalog_auth_token)
+    JSON.parse(response.body).fetch('repositories', [])
+  end
+
+  def catalog_token(system)
+    system = System.where(proxy_byos: false).last if system.nil?
```

Comment:
Why do we not allow systems that use the update server as a proxy? We want to allow any system that can register to the update server to access the registry if that is "allowed" according to what is included in a product.

Reply:
Good catch

That check does not ban access to BYOS system, we do allow them access but, for the catalog/search request, during testing, there was a request made without system, system being nil, and it crashed

That scenario is fixed (login, pull and search operations succeed) and the check should be removed